R/utils.R

```diff
@@ -536,51 +536,98 @@ get_mimetype <- function(filename) {
+            else
+                glue::glue('<link ', glue::glue('href="{tagdata}"'), ' rel="stylesheet">')
     }
     else
       stop(sprintf("Invalid URL supplied in external_stylesheets. Please check the syntax used for this parameter."), call. = FALSE)
   } else {
     # strip leading slash from href if present
+    if (is.list(tagdata))
+      href <- tagdata$src
+    else
+      href <- tagdata    
     href <- sub("^/", "", href)
     modified <- as.integer(file.mtime(local_path))
-    sprintf("<link href=\"%s%s?m=%s\" rel=\"stylesheet\">",
-            prefix,
-            href,
-            modified)
+    glue::glue('<link href="{prefix}{href}?m={modified}" rel="stylesheet">')
   }
 }
 
-generate_js_dist_html <- function(href,
+generate_js_dist_html <- function(tagdata,
                                   local = FALSE,
                                   local_path = NULL,
                                   prefix = NULL,
                                   as_is = FALSE) {
+  attribs <- names(tagdata)
+  allowed_attribs <- c("async",
+                       "crossorigin",
+                       "defer",
+                       "integrity",
+                       "nomodule",
+                       "nonce",
+                       "referrerpolicy",
+                       "src",
+                       "type",
+                       "charset",
+                       "language")
+  if (!all(attribs %in% allowed_attribs)) {
+      stop(sprintf("The following specified script attributes are invalid: ",
+                    paste0(setdiff(attribs, allowed_attribs), collapse=", ")), call. = FALSE)
+  }                                      
   if (!(local)) {
-  if (grepl("^(?:http(s)?:\\/\\/)?[\\w.-]+(?:\\.[\\w\\.-]+)+[\\w\\-\\._~:/?#[\\]@!\\$&'\\(\\)\\*\\+,;=.]+$",
-      href,
-      perl=TRUE) || as_is) {
-      sprintf("<script src=\"%s\"></script>", href)
-    }
+    if (any(grepl("^(?:http(s)?:\\/\\/)?[\\w.-]+(?:\\.[\\w\\.-]+)+[\\w\\-\\._~:/?#[\\]@!\\$&'\\(\\)\\*\\+,;=.]+$",
+        tagdata,
+        perl=TRUE)) || as_is) {
+            if (is.list(tagdata))
+                glue::glue('<script ', glue::glue_collapse(glue::glue('{attribs}="{tagdata}"'), sep=" "), '></script>')
+            else
+                glue::glue('<script ', glue::glue('src="{tagdata}"'), '></script>')
+        }
     else
       stop(sprintf("Invalid URL supplied. Please check the syntax used for this parameter."), call. = FALSE)
   } else {
     # strip leading slash from href if present
+    if (is.list(tagdata))
+      href <- tagdata$src
+    else
+      href <- tagdata
     href <- sub("^/", "", href)
     modified <- as.integer(file.mtime(local_path))
-    sprintf("<script src=\"%s%s?m=%s\"></script>",
-            prefix,
-            href,
-            modified)
+    glue::glue('<script src="{prefix}{href}?m={modified}"></script>')
```

Comment:
Same question as https://github.com/plotly/dashR/pull/226/files#r478448814 -- `href <-` feels like 🍝 from `link`, use `src` instead for clarity.

Reply:
fixed in b2ff7a1